Web-client helper methods

sisou opened the pull request:
## What's in this pull request?

Adds various helper methods to web-client structs that I found are used in our frontend-code, now that I am converting more of the apps.

- `.equals()` and `.compare()` methods for some primitives
- Contract creation/interaction data decoding methods

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

sisou committed:
web-client: Add missing SignatureProof.deserialize and PublicKey.SIZE

sisou committed:
Add helper methods to decode contract recipient data and proofs

And use these helpers in plain transaction encoding.

sisou committed:
web-client: Add equals and compare methods to some primitives

sisou committed:
web-client: simplify some methods

jsdanielh commented on web-client/src/common/address.rs:
```diff
@@ -91,6 +91,16 @@ impl Address {
         self.inner.serialize_to_vec()
     }
+
+    #[cfg(feature = "primitives")]
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/address.rs:
```diff
@@ -91,6 +91,16 @@ impl Address {
+        self.inner == other.inner
+    }
+
+    #[cfg(feature = "primitives")]
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/hashed_time_locked_contract.rs:
```diff
@@ -0,0 +1,132 @@
+use crate::common::transaction::{PlainTransactionProofType, PlainTransactionRecipientDataType};
+
+#[wasm_bindgen]
+pub struct HashedTimeLockedContract;
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/hashed_time_locked_contract.rs:
```diff
@@ -0,0 +1,132 @@
+#[wasm_bindgen]
+impl HashedTimeLockedContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/hashed_time_locked_contract.rs:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/hashed_time_locked_contract.rs:
```diff
@@ -0,0 +1,132 @@
+}
+
+impl HashedTimeLockedContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/hashed_time_locked_contract.rs:
```diff
@@ -0,0 +1,132 @@
+        }))
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/signature_proof.rs:
```diff
@@ -175,16 +180,23 @@ impl SignatureProof {
         self.inner.serialize_to_vec()
     }
+
+    pub fn to_plain(&self) -> Result<PlainTransactionProofType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/signature_proof.rs:
```diff
@@ -240,6 +252,33 @@ impl SignatureProof {
 }
 
+impl SignatureProof {
+    pub fn to_plain_transaction_proof(&self) -> PlainTransactionProof {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/primitives/public_key.rs:
```diff
@@ -24,6 +24,16 @@ impl PublicKey {
 #[wasm_bindgen]
 impl PublicKey {
+    #[wasm_bindgen(getter = SIZE)]
+    pub fn size() -> usize {
```
Could you add rustdoc?

sisou replied:
Self-documenting.

jsdanielh commented on web-client/src/primitives/public_key.rs:
```diff
@@ -24,6 +24,16 @@ impl PublicKey {
+    }
+
+    #[wasm_bindgen(getter = serializedSize)]
+    pub fn serialized_size(&self) -> usize {
```
Could you add rustdoc?

sisou replied:
Self-documenting.

jsdanielh commented on web-client/src/primitives/public_key.rs:
```diff
@@ -101,6 +111,14 @@ impl PublicKey {
         Address::from(nimiq_keys::Address::from(&self.inner))
     }
+
+    pub fn equals(&self, other: &PublicKey) -> bool {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/primitives/public_key.rs:
```diff
@@ -101,6 +111,14 @@ impl PublicKey {
+        self.inner == other.inner
+    }
+
+    pub fn compare(&self, other: &PublicKey) -> i32 {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/primitives/private_key.rs:
```diff
@@ -70,6 +70,10 @@ impl PrivateKey {
         self.inner.to_hex()
     }
+
+    pub fn equals(&self, other: &PrivateKey) -> bool {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/primitives/es256_public_key.rs:
```diff
@@ -118,6 +115,14 @@ impl ES256PublicKey {
         Address::from(nimiq_keys::Address::from(&self.inner))
     }
+
+    pub fn equals(&self, other: &ES256PublicKey) -> bool {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/primitives/es256_public_key.rs:
```diff
@@ -118,6 +115,14 @@ impl ES256PublicKey {
+        self.inner == other.inner
+    }
+
+    pub fn compare(&self, other: &ES256PublicKey) -> i32 {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/transaction.rs:
```diff
@@ -820,24 +610,24 @@ pub struct PlainHtlcData {
-    reward_address: String,
-    signal_data: Option<String>,
-    proof_of_knowledge: String,
+    pub raw: String,
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/transaction.rs:
```diff
@@ -820,24 +610,24 @@ pub struct PlainHtlcData {
-    new_reward_address: Option<String>,
-    new_signal_data: Option<Option<String>>,
-    new_proof_of_knowledge: Option<String>,
+    pub raw: String,
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/vesting_contract.rs:
```diff
@@ -0,0 +1,48 @@
+#[wasm_bindgen]
+impl VestingContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/vesting_contract.rs:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/vesting_contract.rs:
```diff
@@ -0,0 +1,48 @@
+}
+
+impl VestingContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/vesting_contract.rs:
```diff
@@ -0,0 +1,48 @@
+        }))
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/vesting_contract.rs:
```diff
@@ -0,0 +1,48 @@
+};
+
+#[wasm_bindgen]
+pub struct VestingContract;
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/staking_contract.rs:
```diff
@@ -0,0 +1,134 @@
+};
+
+#[wasm_bindgen]
+pub struct StakingContract;
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/staking_contract.rs:
```diff
@@ -0,0 +1,134 @@
+#[wasm_bindgen]
+impl StakingContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/staking_contract.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/staking_contract.rs:
```diff
@@ -0,0 +1,134 @@
+}
+
+impl StakingContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Could you add rustdoc?

jsdanielh commented on web-client/src/common/staking_contract.rs:
```diff
@@ -0,0 +1,134 @@
+        })
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Could you add rustdoc?

sisou committed:
Add rustdoc to public methods

[skip ci]

sisou commented:
@jsdanielh I rebased on latest `albatross` and added rustdocs to all WASM-exposed methods. I resolved those comments above. The remaining ones are methods only used internally between structs, not exposed to userland.
